modules/core/shared/src/main/scala/net/BitVectorSocket.scala

```diff
@@ -48,20 +48,47 @@ object BitVectorSocket {
         case finite: FiniteDuration => _.timeout(finite)
       }
 
-      def readBytes(n: Int): F[Array[Byte]] =
+      def readBytes(n: Int): F[Chunk[Byte]] =
         withTimeout(socket.readN(n)).flatMap { c =>
-          if (c.size == n) c.toArray.pure[F]
+          if (c.size == n) c.pure[F]
           else ev.raiseError(EofException(n, c.size))
         }
 
       override def read(nBytes: Int): F[BitVector] =
-        readBytes(nBytes).map(BitVector(_))
+        readBytes(nBytes).map(_.toByteVector.bits)
 
       override def write(bits: BitVector): F[Unit] =
-        socket.write(Chunk.array(bits.toByteArray))
+        socket.write(Chunk.byteVector(bits.bytes))
 
     }
 
+  def buffered[F[_]](socket: Socket[F], readTimeout: Duration, carryRef: Ref[F, BitVector])(implicit F: Temporal[F]): BitVectorSocket[F] =
+    new BitVectorSocket[F] {
+      val withTimeout: F[Option[Chunk[Byte]]] => F[Option[Chunk[Byte]]] = readTimeout match {
+        case _: Duration.Infinite   => identity
+        case finite: FiniteDuration => _.timeout(finite)
+      }
+
+
+      private def doRead =
+        withTimeout(socket.read(Int.MaxValue)).flatMap { 
+          case Some(bytes) => carryRef.update(_ ++ bytes.toByteVector.bits)
+          case None => F.unit
+        }
+
+      override def read(nBytes: Int): F[BitVector] = {
+        val nBits = nBytes * 8
+        // TODO mutex & avoid double access to carryRef
+        carryRef.get.flatMap { carry =>
```

Comment:
Maybe `carryRef` should be an `AtomicCell` ?

https://typelevel.org/cats-effect/docs/std/atomic-cell

Reply:
Ha, nice! Definitely `AtomicCell` if we need synchronization. Not sure we do in this case though -- adding it makes test about 33% slower. Thinking maybe we should drop `BitVectorSocket` entirely or at least make `MessageSocket` smarter about performing reads -- e.g., build a whole `Decoder[Message]` instead of reading a few bytes, decoding a few bytes, etc.